src/index.tsx

```diff
@@ -1,36 +1,217 @@
-  return !ComponentOrFunction.prototype.isReactComponent;
+export function useResponsiveImageView({
+  aspectRatio: controlledAspectRatio,
+  source: initialSource,
+  onLoad = defaultOnLoad,
+  onError = defaultOnError,
+}: UseResponsiveImageViewOptions): ResponsiveImageViewBag {
+  if (!initialSource) {
+    throw new Error('"source" is required');
+  }
+
+  const [state, dispatch] = React.useReducer(reducer, initialState);
+
+  function retry() {
+    dispatch({ type: 'RETRY' });
+  }
+
+  function isAspectRatioControlled() {
+    return controlledAspectRatio !== undefined;
+  }
+
+  function getAspectRatio() {
+    return isAspectRatioControlled() ? controlledAspectRatio : state.aspectRatio;
+  }
+
+  function getImageProps({
+    source,
+    style = {},
+    ...props
+  }: Parameters<ResponsiveImageViewBag['getImageProps']>[0] = {}) {
+    const imageProps: ImageProps = {
+      source: initialSource,
+      style: [style, { height: '100%', width: '100%' }],
+      ...props,
+    };
+    return imageProps;
+  }
+
+  function getViewProps({
+    style = {},
+    ...props
+  }: Parameters<ResponsiveImageViewBag['getViewProps']>[0] = {}) {
+    return { style: [style, { aspectRatio: getAspectRatio() }], ...props };
+  }
+
+  React.useEffect(() => {
+    let pendingGetImageSize = { cancel: /* istanbul ignore next: just a stub  */ () => {} };
+
+    function handleImageSizeSuccess(width: number, height: number) {
+      onLoad();
+      dispatch({ type: 'SUCCESS', payload: width / height });
+    }
+
+    function handleImageSizeFailure(err: any) {
+      const errMessage =
+        err instanceof Error ? err.message : /* istanbul ignore next */ String(err);
+      onError(errMessage);
+      dispatch({ type: 'FAILURE', payload: errMessage });
+    }
+
+    if (typeof initialSource === 'object' && initialSource.uri) {
+      // Retrieve image dimensions from URI
+      pendingGetImageSize = getImageSize(
+        initialSource.uri,
+        handleImageSizeSuccess,
+        handleImageSizeFailure,
+      );
+    } else {
+      // Retrieve image dimensions from imported resource
+      const imageSource = Image.resolveAssetSource(initialSource);
+      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+      if (imageSource) {
+        handleImageSizeSuccess(imageSource.width, imageSource.height);
+      } else {
+        handleImageSizeFailure(new Error('Failed to retrieve image dimensions.'));
+      }
+    }
+
+    return () => {
+      pendingGetImageSize.cancel();
+    };
+    // Using JSON.stringify here because the `source` parameter can be a nested
+    // object. The alternative is requiring the user to memoize it, by why make
+    // them do that when we don't have to? (Note: they already have to memoize
+    // onLoad and onError, but those are much less likely to be used.)
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [JSON.stringify(initialSource), onLoad, onError, state.retryCount]);
```

Comment:
@coderabbitai Can we do any better here as far as preventing the consumer from having to memoize the options they pass to our hook (assuming they're not using React Compiler)? I was thinking maybe the [latest ref pattern](https://www.epicreact.dev/the-latest-ref-pattern-in-react) for `onLoad` and `onError`, but wasn't sure if that was safe to use in this circumstance. And perhaps there's a better way than JSON.stringify that I don't know about which would prevent them from having to memoize `initialSource`?

Reply:
@coderabbitai Regarding the use of JSON.stringify in src/index.tsx, I agree it's not great, but you seem to just be making things up about sourceWidth and sourceHeight - I don't care about these (if they're even real). `source` (`initialSource`) is either going to be a number for required/imported resources or an object with a `uri` property - those are the two things I care about.